tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,397 @@
+///
+/// ```sh
+/// perf record --call-graph dwarf target/release/mybin
+/// perf script | inferno-collapse-perf > stacks.folded
+/// ```
+///
+/// # Sample Counts
+///
+/// Because `tracing-flame` doesn't use sampling, the number at the end of each
+/// folded stack trace does not represent a number of samples of that stack.
+/// Instead, the numbers on each line are the number of nanoseconds since the
+/// last event in the same thread.
+///
+/// # Dropping and Flushing
+///
+/// Because of the way that `tracing` works, if you use a global subscriber the
+/// drop implementations on your various layers will not get called when your
+/// program exits. This means that if you're using a buffered writer as the
+/// inner writer for the `FlameLayer` you're not guaranteed to see all the
+/// events that have been emitted in the file by default.
+///
+/// To ensure all data is flushed when the program exits, `FlameLayer` exposes
+/// the [`flush_on_drop`] function, which returns a [`FlushGuard`]. The `FlushGuard`
+/// will flush the writer when it is dropped. If necessary, it can also be used to manually
+/// flush the writer.
+///
+/// [`flush_on_drop`]: struct.FlameLayer.html#method.flush_on_drop
+/// [`FlushGuard`]: struct.FlushGuard.html
+#[derive(Debug)]
+pub struct FlameLayer<S, W>
+where
+    S: Subscriber,
+    W: Write + 'static,
+{
+    out: Arc<Mutex<W>>,
+    _inner: PhantomData<S>,
+}
+
+/// An RAII guard for managing flushing a global writer that is
+/// otherwise inaccessible.
+///
+/// This type is only needed when using
+/// `tracing::subscriber::set_global_default`, which prevents the drop
+/// implementation of layers from running when the program exits.
+#[must_use]
+#[derive(Debug)]
+pub struct FlushGuard<W>
+where
+    W: Write + 'static,
+{
+    out: Arc<Mutex<W>>,
+}
+
+impl<S, W> FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
+{
+    /// Return a new FlameLayer that outputs all folded stack samples to the
+    /// provided writer
+    pub fn new(writer: W) -> Self {
+        // Initialize the start used by all threads when initializing the
+        // LAST_EVENT when constructing the layer
+        START.with(|_| ());
+        Self {
+            out: Arc::new(Mutex::new(writer)),
+            _inner: PhantomData,
+        }
+    }
+
+    /// Return a `FlushGuard` which will flush the `FlameLayer`'s writer when
+    /// it is dropped, or when `flush` is manually invoked on the guard.
+    pub fn flush_on_drop(&self) -> FlushGuard<W> {
+        FlushGuard {
+            out: self.out.clone(),
+        }
+    }
+}
+
+impl<W> FlushGuard<W>
+where
+    W: Write + 'static,
+{
+    /// Flush the internal writer of the `FlameLayer`, ensuring that all
+    /// intermediately buffered contents reach their destination.
+    pub fn flush(&self) -> Result<(), Error> {
+        Ok(self
+            .out
+            .lock()
+            .map_err(|_| Kind::LockError)
```

Comment:
I think we should probably just panic if the mutex is poisoned (possibly guarded by a check that the current thread isn't panicking, in case users call this in `Drop` impls.